seahub/file_participants/api.py

```diff
@@ -0,0 +1,149 @@
+        """
+        # argument check
+        path = request.GET.get('path')
+        if not path:
+            return api_error(status.HTTP_400_BAD_REQUEST, 'path invalid.')
+        path = normalize_file_path(path)
+
+        # resource check
+        try:
+            file_id = seafile_api.get_file_id_by_path(repo_id, path)
+        except SearpcError as e:
+            logger.error(e)
+            return api_error(status.HTTP_500_INTERNAL_SERVER_ERROR, 'Internal Server Error.')
+        if not file_id:
+            return api_error(status.HTTP_404_NOT_FOUND, 'File %s not found.' % path)
+
+        # permission check
+        if not check_folder_permission(request, repo_id, '/'):
+            return api_error(status.HTTP_403_FORBIDDEN, 'Permission denied.')
+
+        # main
+        file_uuid = FileUUIDMap.objects.get_or_create_file_uuid_by_path(repo_id, path)
+
+        participant_list = []
+        participant_queryset = FileParticipant.objects.get_participants(file_uuid)
+
+        for participant in participant_queryset:
+            participant_info = get_user_common_info(participant.username)
+            participant_list.append(participant_info)
+
+        return Response({'participant_list': participant_list})
+
+    def post(self, request, repo_id, format=None):
+        """Post a participant of a file.
+        """
+        # argument check
+        path = request.data.get('path')
+        if not path:
+            return api_error(status.HTTP_400_BAD_REQUEST, 'path invalid.')
+        path = normalize_file_path(path)
+
+        email = request.data.get('email', '').lower()
+        if not email or not is_valid_username(email):
+            return api_error(status.HTTP_400_BAD_REQUEST, 'email invalid.')
+
+        # resource check
+        try:
+            file_id = seafile_api.get_file_id_by_path(repo_id, path)
+        except SearpcError as e:
+            logger.error(e)
+            return api_error(status.HTTP_500_INTERNAL_SERVER_ERROR, 'Internal Server Error.')
+        if not file_id:
+            return api_error(status.HTTP_404_NOT_FOUND, 'File %s not found.' % path)
+
+        try:
+            user = User.objects.get(email=email)
+        except User.DoesNotExist:
+            return api_error(status.HTTP_404_NOT_FOUND, 'User %s not found.' % email)
+
+        # permission check
+        if not check_folder_permission(request, repo_id, '/'):
+            return api_error(status.HTTP_403_FORBIDDEN, 'Permission denied.')
+
+        request.user = user
+        if not check_folder_permission(request, repo_id, '/'):
+            return api_error(status.HTTP_403_FORBIDDEN, _('%s Permission denied.') % email)
+
+        # main
+        file_uuid = FileUUIDMap.objects.get_or_create_file_uuid_by_path(repo_id, path)
+
+        if FileParticipant.objects.get_participant(file_uuid, email):
+            return api_error(status.HTTP_400_BAD_REQUEST, _('Participant %s already exists.') % email)
```

Comment:
The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. The response body SHOULD include enough

information for the user to recognize the source of the conflict. Ideally, the response entity would include enough information for the user or user agent to fix the problem; however, that might not be possible and is not required.

Conflicts are most likely to occur in response to a PUT request. For example, if versioning were being used and the entity being PUT included changes to a resource which conflict with those made by an earlier (third-party) request, the server might use the 409 response to indicate that it can't complete the request. In this case, the response entity would likely contain a list of the differences between the two versions in a format defined by the response Content-Type.